frame/executive/src/lib.rs

```diff
@@ -349,12 +357,39 @@ where
 	/// Execute given extrinsics and take care of post-extrinsics book-keeping.
 	fn execute_extrinsics_with_book_keeping(
```

Comment:
block producer will not use execute_extrinsics_with_book_keeping when building the block, instead they the use the runtime API: BlockBuilder::apply_extrinsic which calls Executive::apply_extrinsic, and BlockBuilder::finalize_block
So the detection of the first non-inherent should be done in apply_extrinsic and execute the post_inherent hook before the first non-inherent.
Also if there was no non-inherent extrinsic, the finalize_block should execute post_inherent hook

Reply:
maybe you can do something similar to EXTRINSIC_INDEX in frame_system: a storage where you put if the previous was an inherent or not.